examples/channelscan/main.go

```diff
@@ -0,0 +1,72 @@
+// This example program shows using Go routines and channels to coordinate
+// BLE scanning.
+//
+// The first Go routine starts scanning using the BLE adaptor. When it finds
+// a new device, it puts the information into a channel so it can be displayed.
+//
+// The second Go routine is a ticker that puts a "true" value into a channel every 3 seconds.
+//
+// The main function uses a select{} statement to wait until one of the two channels is unblocked
+// by receiving data. If a new device is found, the boolean variable named "found" will
+// be set to true, so that the timeout is reset for each 3 second period.
+//
+package main
+
+import (
+	"time"
+
+	"tinygo.org/x/bluetooth"
+)
+
+var (
+	adapter = bluetooth.DefaultAdapter
+	devices = make(chan *bluetooth.ScanResult, 1)
+	ticker  = make(chan bool, 1)
+	found   = true
+)
+
+func main() {
+	// Enable BLE interface.
+	if err := adapter.Enable(); err != nil {
+		panic("failed to enable adaptor:" + err.Error())
+	}
+
+	// Start scanning
+	go performScan()
+
+	// Start timeout ticker
+	go startTicker()
+
+	// Wait for devices to be scanned
+	for {
+		select {
+		case device := <-devices:
+			found = true
+			println("found device:", device.Address.String(), device.RSSI, device.LocalName())
+		case <-ticker:
+			if !found {
+				println("no devices found in last 3 seconds...")
+			}
+			found = false
+		}
+	}
+}
+
+func performScan() {
+	println("scanning...")
+
+	err := adapter.Scan(func(adapter *bluetooth.Adapter, device bluetooth.ScanResult) {
+		devices <- &device
```

Comment:
This will do a heap allocation in an interrupt, which is unsafe.

There may not be a safe way to do this. In fact, I suspect the existing scan code is unsafe regarding heap allocations in interrupts.

Reply:
So how should this be handled exactly in this kind of scenario?

Reply:
(Just re-discovering this PR)

You can send the plain `device` struct over the channel, like this:

```go
select {
case devices <- device:
default:
}
```

And change `devices` to not use a pointer: `devices = make(chan bluetooth.ScanResult, 1)`.

Reply:
I was just ahead of my time :)

Will rebase and make suitable corrections.

Reply:
OK done.